### PR TITLE
Send to all urns

### PR DESCRIPTION
**Pull request by norkans7**

### CODE REVIEW

**Comment by nicpottier** [comment]
coverage!

**Comment by norkans7** [comment]
I am adding more tests

**Comment by norkans7** [comment]
Addresses https://github.com/rapidpro/rapidpro/issues/451

File: temba/flows/models.py [hunk 1]
```diff
@@ -5128,22 +5132,46 @@ def execute(self, run, actionset_uuid, msg, offline_on=None):
...
                 context = run.flow.build_message_context(run.contact, msg)
 
+                reply_msgs = None
                 try:
                     if msg:
-                        reply = msg.reply(text, user, trigger_send=False, message_context=context, session=run.session,
-                                          media=media, msg_type=self.MSG_TYPE)
+                        reply_msgs = msg.reply(text, user, trigger_send=False, message_context=context,
+                                               session=run.session, msg_type=self.MSG_TYPE, media=media,
+                                               send_all=self.send_all)
```
Comment (ericnewcomer) [inline temba/flows/models.py]: Why is the implementation so different for the offline / online case? Seems they could use the same method.

File: templates/partials/action_editor.haml [hunk 2]
```diff
@@ -176,6 +176,15 @@
...
 .modal-footer
+
+  .send-all{ng-show:'config.type == "reply"', style:'float:left'}
+
+    %label
+      %input{type:'checkbox', ng-model:'action.send_all'}
+
+      %span
+        -trans "Send to all URNs"
+
```
Comment (ericnewcomer) [inline templates/partials/action_editor.haml]: screenshot?

Comment (norkans7) [inline templates/partials/action_editor.haml]: send all

File: temba/flows/models.py (hunk 1 above)
Comment (norkans7) [inline temba/flows/models.py]: offline case create message with the status as SENT, and yes I see now I can use the same path, I change this

File: templates/partials/action_editor.haml (hunk 2 above)
Comment (ericnewcomer) [inline templates/partials/action_editor.haml]: How about "Send to all contact addresses"

Comment (norkans7) [inline templates/partials/action_editor.haml]: yep

File: temba/msgs/models.py [hunk 3]
```diff
@@ -224,10 +224,13 @@ class Broadcast(models.Model):
     media_dict = models.TextField(verbose_name=_("Media"),
                                   help_text=_("The localized versions of the media"), null=True)
 
+    send_all = models.BooleanField(default=False,
```
Comment (rowanseymour) [inline temba/msgs/models.py]: Won't this be slow to add on a prod database? I think we have to do the three step add here, first with `null=True`, then populate, then `null=False`

File: temba/contacts/models.py [hunk 4]
```diff
@@ -1794,17 +1794,49 @@ def raw_tel(self):
...
+                                  message_context=message_context, session=session, media=media,
+                                  msg_type=msg_type or INBOX, status=status, created_on=created_on)
 
         if trigger_send:
             self.org.trigger_send([msg])
 
         return msg
 
+    def send_all(self, text, user, trigger_send=True, response_to=None, message_context=None, session=None, media=None,
+                 msg_type=None, status='P', created_on=None):
```
Comment (rowanseymour) [inline temba/contacts/models.py]: Not keen on passing status here - I feel like `ReplyAction` shouldn't be concerned about message statuses.

Can't we use `created_on != None` as way of determining whether we're online or offline ?

File: temba/flows/models.py [hunk 5]
```diff
@@ -5121,21 +5125,40 @@ def execute(self, run, context, actionset_uuid, msg, offline_on=None):
...
-            else:
-                try:
-                    if msg:
-                        reply = msg.reply(text, user, trigger_send=False, message_context=context, session=run.session,
-                                          media=media, msg_type=self.MSG_TYPE)
+                status = SENT
+                context = None
+                created_on = offline_on
+
+            reply_msgs = None
```
Comment (rowanseymour) [inline temba/flows/models.py]: It's confusing to use a variable as either a single object or a list - especially when the name of the variable suggests it's a list

I think you can just use `replies` from above and append messages to it, e.g. 

```
replies.append(msg.reply(...))
```

File: templates/partials/action_editor.haml [hunk 6]
```diff
@@ -176,6 +176,15 @@
         -trans "Ok"
     </ng-form>
 .modal-footer
+
+  .send-all{ng-show:'config.type == "reply"', style:'float:left'}
```
Comment (rowanseymour) [inline templates/partials/action_editor.haml]: Screenshot?

File: temba/flows/models.py [hunk 7]
```diff
@@ -5121,21 +5125,40 @@ def execute(self, run, context, actionset_uuid, msg, offline_on=None):
                 if media_url:
                     media = "%s:https://%s/%s" % (media_type, settings.AWS_BUCKET_DOMAIN, media_url)
 
+            status = PENDING
```
Comment (rowanseymour) [inline temba/flows/models.py]: General stylistic comment
```
a  = 1
if something:
    a = 2
```

is clearer as

```
if something:
    a = 2
else:
    a = 1
```

so you're not assigning and then immediately reassigning a variable

File: templates/partials/action_editor.haml (hunk 6 above)
Comment (norkans7) [inline templates/partials/action_editor.haml]: send-all

File: temba/msgs/models.py (hunk 3 above)
Comment (norkans7) [inline temba/msgs/models.py]: I think default=False is fine here, I tried the 3 steps and I am getting a suggestion to use NullBooleanField

Comment (rowanseymour) [inline temba/msgs/models.py]: AFAIK that's gonna do a table lock on `msgs_broadcast` while it writes millions of FALSE values. Did you try this migration on a large database?

File: temba/contacts/models.py [hunk 8]
```diff
@@ -1794,17 +1794,59 @@ def raw_tel(self):
...
 
-    def send(self, text, user, trigger_send=True, response_to=None, message_context=None, session=None, media=None, msg_type=None):
+    def send(self, text, user, trigger_send=True, response_to=None, message_context=None, session=None, media=None,
+             msg_type=None, created_on=None):
         from temba.msgs.models import Msg, INBOX
 
-        msg = Msg.create_outgoing(self.org, user, self, text, priority=Msg.PRIORITY_HIGH, response_to=response_to,
-                                  message_context=message_context, session=session, media=media, msg_type=msg_type or INBOX)
+        if created_on is None:
+            status = 'P'
```
Comment (rowanseymour) [inline temba/contacts/models.py]: These can be constants - you can import above

File: temba/contacts/models.py [hunk 9]
```diff
@@ -1794,17 +1794,59 @@ def raw_tel(self):
...
+        else:
+            status = 'S'
+
+        contact_urns = self.get_urns()
+        for c_urn in contact_urns:
+            try:
+
+                recipient = c_urn
+                if status != 'P':
+                    recipient = (c_urn.contact, c_urn)
```
Comment (rowanseymour) [inline temba/contacts/models.py]: I think recipient here should just be the URN, no?

File: temba/flows/models.py [hunk 10]
```diff
@@ -5122,20 +5127,35 @@ def execute(self, run, context, actionset_uuid, msg, offline_on=None):
...
-                                          media=media, msg_type=self.MSG_TYPE)
+                created_on = None
+
+            try:
+                if msg:
+                    reply_msg = msg.reply(text, user, trigger_send=False, message_context=context,
+                                          session=run.session, msg_type=self.MSG_TYPE, media=media,
+                                          send_all=self.send_all, created_on=created_on)
+
+                    if reply_msg:
```
Comment (rowanseymour) [inline temba/flows/models.py]: Don't think `Msg.reply` can return None

File: temba/flows/models.py [hunk 11]
```diff
@@ -5122,20 +5127,35 @@ def execute(self, run, context, actionset_uuid, msg, offline_on=None):
...
+                                                       created_on=created_on)
                     else:
-                        reply = run.contact.send(text, user, trigger_send=False, message_context=context,
-                                                 session=run.session, msg_type=self.MSG_TYPE, media=media)
-                except UnreachableException:
-                    pass
+                        reply_msg = run.contact.send(text, user, trigger_send=False, message_context=context,
+                                                     session=run.session, msg_type=self.MSG_TYPE, media=media,
+                                                     created_on=created_on)
+                        if reply_msg:
```
Comment (rowanseymour) [inline temba/flows/models.py]: Don't think `Contact.send` can return None

Comment (norkans7) [inline temba/flows/models.py]: it is possible in case of loops

Comment (rowanseymour) [inline temba/flows/models.py]: Oh right, n/m

File: temba/msgs/migrations/0087_populate_broadcast_send_all.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.6 on 2017-04-04 14:57
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+def apply_as_migration(apps, schema_editor):
+    Broadcast = apps.get_model('msgs', 'Broadcast')
+    Broadcast.objects.all().update(send_all=False)
```
Comment (rowanseymour) [inline temba/msgs/migrations/0087_populate_broadcast_send_all.py]: This needs to be batched to avoid a really long lock on `msgs_broadcast`

**Comment by norkans7** [comment]
Replaced by #1215 #1216 #1217

**Pull request closed without merging**